### PR TITLE
Feat/cube exchange connector

### PR DESCRIPTION
**Pull request by mlguys**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
- Spot Connector for Cube Exchange: https://www.cube.exchange/trade/SOLUSDC
- Standard Spot Connector
- Supported Live and Staging environment
- Included Rate Oracle for Cube Exchange

**Tests performed by the developer**:
- Standard unit tests

**Tips for QA testing**:
- Standard spot connector QA testing
- To test on staging, please sign-up account on https://dev.cube.exchange then please contact me for approving the KYC process.

### CODE REVIEW

**Comment by MHHukiewitz** [comment]
Docs have been updated here: https://github.com/hummingbot/hummingbot-site/pull/358

**Comment by rapcmia** [comment]
PR update:
- Create live and staging account ✅ 
- When adding invalid API keys:
   image
    - When selecting cube environment, it still accepts if not `live` or `staging`
    - However this can be corrected if you run `connector cube` again
    - When creating API, make sure which account is selected and select `write` permission
 - Ran tests on staging, funds sent by developer ✅ 
- Test v1 pmm
    - Orders created and cancelled accordingly
    - Orders are created successfully on client and exchange
    - Ran tests on default setup and 180 refresh time, no errors found
 - Test v2 pmm_simple
    - Orders created ok
    - Unable to send market order
      ```
      2024-04-08 22:18:27,626 - 9728 - hummingbot.connector.exchange.cube.cube_exchange.CubeExchange - NETWORK - Error submitting sell MARKET order to Cube for 0.4159 TSOL-TUSDC NaN.
      Traceback (most recent call last):
        File "/Users/rapcomia/github/hummingbot/6830/hummingbot/connector/exchange_py_base.py", line 452, in _create_order
          await self._place_order_and_process_update(order=order, **kwargs,)
        File "/Users/rapcomia/github/hummingbot/6830/hummingbot/connector/exchange_py_base.py", line 469, in _place_order_and_process_update
          exchange_order_id, update_timestamp = await self._place_order(
        File "/Users/rapcomia/github/hummingbot/6830/hummingbot/connector/exchange/cube/cube_exchange.py", line 221, in _place_order
          "price": int(round(exchange_price)),
      ValueError: cannot convert NaN to integer
      ```
       - See attach files [04082024.zip](https://github.com/hummingbot/hummingbot/files/14906738/04082024.zip)

---

Pending
- Run test on mainnet

**Comment by mlguys** [comment]
Thanks for the testing @rapcmia, will push fix soon.

**Comment by MHHukiewitz** [comment]
Just one thought about the staging config: Will regular users be able to access the staging environment? I don't think so. In that case, it will only cause extra work/confusion for the user. Could we hide this config in the config file and not ask for it in the UI, so that we assume `live` by default?

**Comment by mlguys** [comment]
@MHHukiewitz: yes we can hide this setting and assume `live` as default.

**Comment by rapcmia** [comment]
PR update:
- Create orders and cancel ok
- Ran tests on long and short refresh time
- Filled orders recorded ok
- Ran multiple levels ok 
- Ran market order script, same error as reported on `staging`
   - [gist](https://gist.github.com/rapcmia/eab4414b5b62a9607db8e6c977945bf6) of market order script used on the test

---
Pending
- Skipped testing v2 strategy due to error found on market orders

**Comment by mlguys** [comment]
@rapcmia Added fix for the issue on market order getting error

**Review by rapcmia — APPROVED** [review]
- Create staging and mainnet account ok 
- Connect API successfully, balance displayed ok 
	- Invalid keys return `unhandled error` user must restart the hummingbot client 
- Ran tests on both mainnet and testnet ok 
- Ran tests on Mainnet on both v1 and v2 (pmm_single_level)
	- Create orders ok, matched between PR and exchange 
	- Filled orders recorded ok on CSV and exchange history
	- Market order supported for v2 strategy
	- Ran tests on long and short refresh time ok 
- Ran tests on rate oracle cube
	- Ran -p on markets on both mainnet and testnet
	- initial balance would display 0 therefore user must change global_token to USDC or TUSDC(tsetnet)
- Build docker image and run same tests, all ok

**Comment by mlguys** [comment]
@rapcmia added validators to handle this issue:

> Invalid keys return unhandled error user must restart the hummingbot client
image

**Review by cardosofede — APPROVED** [review]
All good for me, just check the comment on the protobuf file

File: hummingbot/connector/exchange/cube/cube_ws_protobufs/trade_pb2.pyi [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+    __slots__ = ()
+    CANCEL_RESTING: _ClassVar[SelfTradePrevention]
+    CANCEL_AGGRESSING: _ClassVar[SelfTradePrevention]
+    ALLOW_SELF_TRADE: _ClassVar[SelfTradePrevention]
+
+class PostOnly(int, metaclass=_enum_type_wrapper.EnumTypeWrapper):
+    __slots__ = ()
+    DISABLED: _ClassVar[PostOnly]
+    ENABLED: _ClassVar[PostOnly]
+BID: Side
```
[the remark this answers is not in the document]
Comment (mlguys) [reply hummingbot/connector/exchange/cube/cube_ws_protobufs/trade_pb2.pyi:36]: I think so since these files are auto generated from protobuf schema.

**Comment by rapcmia** [comment]
Merged to development and part of release version 1.27.0

**Pull request merged**